Vagrant setup (#76)
tinkerbell/tinkerbell.org: content/setup/local-with-vagrant/_index.md

+++
title = "Local with Vagrant"
date = 2020-06-08T16:16:15+05:30
draft = false
weight = 20
toc = true
+++

An easy way to try out Tinkerbell locally is via [Vagrant](https://www.vagrantup.com).

Vagrant wraps and glues together different technologies to help you write as
code repetitive environment mixing virtualization or cloud providers such as
VirtualBox, libvirt, AWS, GCP with provisioning techniques like Puppet, Ansible,
Bash.

In this context we use VirtualBox or libvirt as a virtualization platform and
bash for provisioning all the required tools and configurations.

Following this tutorial, you will end up with a provisioner up and running and a
worker ready to handle actions and workflows.

## Prerequisites

- [The host's processor should support virtualization](https://www.cyberciti.biz/faq/linux-xen-vmware-kvm-intel-vt-amd-v-support/)
- [Vagrant](https://www.vagrantup.com/downloads) is installed

> - Either [VirtualBox](https://www.virtualbox.org/) or [libvirtd](https://libvirt.org/) is installed and configured.

Reviewer: Isn't libvirtd for Linux users? Can we make it clearer?

Author: In theory you can run both in both platforms. But I never tried libvirt https://libvirt.org/platforms.html

   vagrant configuration (Vagrantfile) and the scripts needed to provide workers
   and provisioners.

```
$ git clone https://github.com/tinkerbell/tink.git
$ cd tink/deploy/vagrant
```

## Start the provisioner

```
$ vagrant up provisioner
Bringing machine 'provisioner' up with 'virtualbox' provider...

    ......

    INFO: tinkerbell stack setup completed successfully on ubuntu server
    NEXT: 1. Enter ./deploy and run: source ../envrc; docker-compose up
           2. Try executing your fist workflow.
              Follow the steps described in https://tinkerbell.org/examples/hello-world/ to say 'Hello World!' with a workflow.
```

When the provisioner is ready, a summary is printed to the console and you can now ssh in.

### Connecting to the Provisioner

All the following commands have to be executed in the artifacts' directory created above unless stated otherwise.

Connect to the provisioner via `vagrant ssh`:

```
$ vagrant ssh provisioner
vagrant@provisioner:~$
```

At this point, you are in an Ubuntu box that has a couple of utils installed.

Let's now start up the Tinkerbell stack with `docker-compose`:

```
$ sudo su -l

> $ cd /vagrant && source envrc && cd deploy

Reviewer: Since you ran `sudo`. Now prefix commands with `#` to show we're running as root?

```
vagrant@provisioner:~$ sudo su -l
root@provisioner:~# cd /vagrant && source envrc && cd deploy
root@provisioner:/vagrant/deploy# docker-compose up -d
```

Author: This is not required as sudo anymore!! 👍 didn't know about the `#` thing at all.
Thanks

-------------------------------------------------------------------------------------------------------------------------
deploy_boots_1         /boots -dhcp-addr 0.0.0.0: ...   Up
deploy_cacher_1        /cacher                          Up             0.0.0.0:42111->42111/tcp, 0.0.0.0:42112->42112/tcp
deploy_db_1            docker-entrypoint.sh postgres    Up (healthy)   0.0.0.0:5432->5432/tcp
deploy_hegel_1         cmd/hegel                        Up
deploy_nginx_1         /docker-entrypoint.sh ngin ...   Up             192.168.1.2:80->80/tcp
deploy_registry_1      /entrypoint.sh /etc/docker ...   Up (healthy)
deploy_tink-cli_1      /bin/sh -c sleep infinity        Up
deploy_tink-server_1   tink-server                      Up (healthy)   0.0.0.0:42113->42113/tcp, 0.0.0.0:42114->42114/tcp
```

### Register the worker's hardware

The provisioner is running we can follow the [example called "Hello
world"](/examples/hello-world), I will contextualize it for our vagrant setup
here.

SSH in the provisioner again and we have to define our hello world template and load it to Tinkerbell:

```

> $ cat > hello-world.yml

Reviewer: If you're using `cat >` then you really need to use the EOF syntax.

https://stackoverflow.com/questions/2500436/how-does-cat-eof-work-in-bash

Regular users won't know to hit control + d.

version: "0.1"
name: hello_world_workflow
global_timeout: 600
tasks:
  - name: "hello world"
    worker: "{{.device_1}}"
    actions:
      - name: "hello_world"
        image: hello-world
        timeout: 60
$ docker exec -i deploy_tink-cli_1 tink template create --name hello-world < ./hello-world.yml
Created Template:  75ab8483-6f42-42a9-a80d-a9f6196130df
```

Register the worker with Tinkerbell

```
$ cat > hardware-data.json
{
  "id": "ce2e62ed-826f-4485-a39f-a82bb74338e2",
  "arch": "x86_64",
  "allow_pxe": true,
  "allow_workflow": true,
  "facility_code": "onprem",
  "ip_addresses": [
    {
      "address": "192.168.1.5",
      "address_family": 4,
      "enabled": true,
      "gateway": "192.168.1.1",
      "management": true,
      "netmask": "255.255.255.248",
      "public": false
    }
  ],
  "network_ports": [
    {
      "data": {

>         "mac": "08:00:27:00:00:01"

Reviewer: Is this Mac address being hard-coded? How do you know that it's correct? Did it get encoded in the Vagrantfile?

      },
      "name": "eth0",
      "type": "data"
    }
  ]
}

$ docker exec -i deploy_tink-cli_1 tink hardware push < ./hardware-data.json
2020/06/17 14:12:45 Hardware data pushed successfully
```

Now that we have a template and hardware registered we can trigger the workflow for that particular hardware but first a bit of context on what we just did:

1. A template is used to create workflows. Template are important because you
   can create as many workflows you need from a single template.
2. A workflow needs a target (a hardware) in vagrant it is the worker and it is
   identified via MAC address.

> Let's start the workflow where `-t` is the template ID you got when registering

Reviewer: This needs to be clearer - it's the only part which isn't hard-coded, so when following I copy and pasted your instructions and it didn't work for me. @gianarb

your template and `-r` contains the variables to replace in the template, in
this case only the worker's MAC address is required:

```

> $ docker exec -i deploy_tink-cli_1 tink workflow create \

Reviewer: Same note as above. Perhaps mention the hard-coded Mac earlier on in the doc.

    -t 75ab8483-6f42-42a9-a80d-a9f6196130df \
    -r '{"device_1":"08:00:27:00:00:01"}'
Created Workflow:  a8984b09-566d-47ba-b6c5-fbe482d8ad7f
```

> Note: this MAC address it hard coded in the Vagrantfile.

We need to spin up our first worker now.

## Start a worker

Open a new terminal and get inside the `tink/deploy/vagrant`.

For a worker the MAC address works as a unique identifier, with Vagrant we are
able to set it to fixed value, in this case `080027000001`.

As you did for the provisioner start the worker via vagrant:

```
$ vagrant up worker
```

If using VirtualBox, the worker shows up in a UI and if everything is right you
should see at the end of the boot a login to a custom Alpine that Tinkerbell
netboot. The operating system runs on RAM, the changes won't be persisted across
reboots. You can login with the username `root` and no password is required.

> ![Screenshot from the worker](/images/vagrant-setup-vbox-worker.png)

Reviewer: Perhaps show the user how to enlarge the size of the console? For MacOS users with Retina or 4k screens, it appears as an unreadable size:

<img width="472" alt="Screenshot 2020-06-18 at 12 04 25" src="https://user-images.githubusercontent.com/6358735/85013922-496a4e80-b15d-11ea-8d70-52a26d5f98b3.png">

<img width="1192" alt="Screenshot 2020-06-18 at 12 04 40" src="https://user-images.githubusercontent.com/6358735/85013912-45d6c780-b15d-11ea-98ff-3d8a074d9f04.png">

<img width="763" alt="Screenshot 2020-06-18 at 12 04 33" src="https://user-images.githubusercontent.com/6358735/85013920-48d1b800-b15d-11ea-95f6-789ee1f85c8e.png">

> That's it! Let us know what you think about it on [Slack](/community-slack).

Reviewer: For me this finishes too early. Yes I brought up the worker, but it gave no output about being provisioned

My suggestion for completing the story:

Run `tink-cli` with `workflow` and with `events` to show that hello-world actually executed.

Mention also that they have to create a new workflow for each boot of the machine. 

Also show them how to get logs, this is also missing but a key piece of using Tinkerbell.

I assume that would be the following - with a brief explanation on the debug data that you can find in each one.

```
docker-compose logs tink-server
docker-compose logs boots
docker-compose logs nginx
```

Reviewer: For example:

<img width="2542" alt="Screenshot 2020-06-18 at 12 20 36" src="https://user-images.githubusercontent.com/6358735/85014409-212f1f80-b15e-11ea-8643-fd940c4d919a.png">

And you can see that the workflow (following the tutorial) didn't actually work / execute since there are no events. 

<img width="2346" alt="Screenshot 2020-06-18 at 12 21 34" src="https://user-images.githubusercontent.com/6358735/85014506-4e7bcd80-b15e-11ea-827e-e54ac57e6d38.png">

Author: Hello! Yeah, I am sure there are a lot more we can explain. We are not probably the unique two with opinions here 💯 I presume we will figure out what to add to this documentation next week.

Thanks for your screenshots and tips

+++
title = "Local with Vagrant"
date = 2020-06-08T16:16:15+05:30
draft = false
weight = 20
toc = true
+++

An easy way to try out Tinkerbell locally is via [Vagrant](https://www.vagrantup.com).

Vagrant wraps and glues together different technologies to help you write as
code repetitive environment mixing virtualization or cloud providers such as
VirtualBox, libvirt, AWS, GCP with provisioning techniques like Puppet, Ansible,
Bash.

In this context we use VirtualBox or libvirt as a virtualization platform and
bash for provisioning all the required tools and configurations.

Following this tutorial, you will end up with a provisioner up and running and a
worker ready to handle actions and workflows.

> ## Prerequisite

Reviewer: Prerequisites

> - The host's processor should support virtualization

Reviewer: Tell them how to find out or link to where to learn.

Reviewer 2: Do you have some good material on how to do this? `grep -o 'vmx\|svm' /proc/cpuinfo` works for Linux, but I'm not sure about macOS.

Reviewer: I would suggest Googling it and linking off to that.

Reviewer: (because that's all that is going to happen when a user sees this vague note)

    provisioner: this with a newly generated keypair for better security.
    provisioner:
    provisioner: Inserting generated public key within guest...
    provisioner: Removing insecure key from the guest if it's present...
    provisioner: Key inserted! Disconnecting and reconnecting using new SSH key...
==> provisioner: Machine booted and ready!
==> provisioner: Checking for guest additions in VM...
==> provisioner: Setting hostname...
==> provisioner: Configuring and enabling network interfaces...
==> provisioner: Rsyncing folder: /Users/name/git/tink/ => /vagrant
==> provisioner: Running provisioner: shell...
    provisioner: Running: /var/folders/0_/841vr1ms57s65r7h1_2vm69m0000gn/T/vagrant-shell20200617-5179-181phru.sh
    provisioner: + whoami
    provisioner: root
    provisioner: + cd /vagrant
    provisioner: + main
    provisioner: + export DEBIAN_FRONTEND=noninteractive
    provisioner: + DEBIAN_FRONTEND=noninteractive
    provisioner: + apt-get update
    provisioner: Hit:1 http://us.archive.ubuntu.com/ubuntu bionic InRelease
    provisioner: Get:2 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]

>     provisioner: Get:3 http://us.archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]

Reviewer: Do you need the whole output here or would an abridged version serve the same purpose and make the doc shorter?

Reviewer 2: Heavily trimmed!

    Login Succeeded
    latest: Pulling from tinkerbell/tink-worker
    Digest: sha256:55c99b7cfafd28244fd3f7adbe46d94ada305f0f28553f95dfd241c168f0439e
    Status: Image is up to date for quay.io/tinkerbell/tink-worker:latest
    quay.io/tinkerbell/tink-worker:latest
    The push refers to repository [192.168.1.1/tink-worker]
    c83ceae18bd4: Layer already exists
    a1872780b5d4: Layer already exists
    3e207b409db3: Layer already exists
    latest: digest: sha256:55c99b7cfafd28244fd3f7adbe46d94ada305f0f28553f95dfd241c168f0439e size: 949
    INFO: tinkerbell stack setup completed successfully on ubuntu server
    NEXT: 1. Enter ./deploy and run: source ../envrc; docker-compose up
           2. Try executing your fist workflow.
              Follow the steps described in https://tinkerbell.org/examples/hello-world/ to say 'Hello World!' with a workflow.
```

When the provisioner is ready, a summary is printed to the console and you can now ssh in.

### Connecting to the Provisioner

All the following commands have to be executed in the artifacts' directory created above unless stated otherwise.

> Connect to the provisioner via ssh, vagrant has an until:

Reviewer: Has an until? Has a utility?

```
$ vagrant ssh provisioner
vagrant@provisioner:~$
```

At this point, you are in an Ubuntu box that has a couple of utils installed and
it is not time to start the services to make it a proper Tinkerbell provisioner

```
$ sudo su -l

> $ cd /vagrant && source envrc && cd deploy

Reviewer: Looks good. How do I restart this though? Same instructions or do they differ? Let's say I came back to this a week later after starting it and after a reboot of the computer?

Reviewer 2: It is just any other docker-compose project. Do we need to teach docker compose here, or is a statement that it is a standard docker-compose project sufficient?

Reviewer 2: I added a note saying it is a standard docker-compose project.

$ docker-compose up -d
```

Those commands give you a fully working provisioner that is ready to receive

> templates and workflow targeting workers. You can check that running:

Reviewer: You can check that it is running

        Name                      Command                  State                             Ports
-------------------------------------------------------------------------------------------------------------------------
deploy_boots_1         /boots -dhcp-addr 0.0.0.0: ...   Up
deploy_cacher_1        /cacher                          Up             0.0.0.0:42111->42111/tcp, 0.0.0.0:42112->42112/tcp
deploy_db_1            docker-entrypoint.sh postgres    Up (healthy)   0.0.0.0:5432->5432/tcp
deploy_hegel_1         cmd/hegel                        Up
deploy_nginx_1         /docker-entrypoint.sh ngin ...   Up             192.168.1.2:80->80/tcp
deploy_registry_1      /entrypoint.sh /etc/docker ...   Up (healthy)
deploy_tink-cli_1      /bin/sh -c sleep infinity        Up
deploy_tink-server_1   tink-server                      Up (healthy)   0.0.0.0:42113->42113/tcp, 0.0.0.0:42114->42114/tcp
```

### Register the worker's hardware

The provisioner is running we can follow the [example called "Hello
world"](/examples/hello-world), I will contextualize it for our vagrant setup
here.

SSH in the provisioner again and we have to define our hello world template and load it to Tinkerbell:

> ```

Reviewer: Do we have an official example of a real Operating System yet that we can link to?

Reviewer 2: Not yet, no.

Author: Nop

  "arch": "x86_64",
  "allow_pxe": true,
  "allow_workflow": true,
  "facility_code": "onprem",
  "ip_addresses": [
    {
      "address": "192.168.1.5",
      "address_family": 4,
      "enabled": true,
      "gateway": "192.168.1.1",
      "management": true,
      "netmask": "255.255.255.248",
      "public": false
    }
  ],
  "network_ports": [
    {
      "data": {
        "mac": "08:00:27:00:00:01"
      },
      "name": "eth0",
      "type": "data"
    }
  ]
}

$ docker exec -i deploy_tink-cli_1 tink hardware push < ./hardware-data.json
2020/06/17 14:12:45 Hardware data pushed successfully
```

Now that we have a template and hardware registered we can trigger the workflow for that particular hardware but first a bit of context on what we just did:

1. A template is used to create workflows. Template are important because you
   can create as many workflows you need from a single template.
2. A workflow needs a target (a hardware) in vagrant it is the worker and it is
   identified via mac address.

Let's start the workflow where `-t` is the template ID you got when registering
your template and `-r` contains the variables to replace in the template, in

> this case only the worker's mac address is required:

Reviewer: Explain how to fetch that address, or if you'd hardcoded it, explain that for the user.

Reviewer 2: Good idea, fixup'd.

worker ready to handle actions and workflows.

## Prerequisite

A bunch of information you have to know and tools you have to install in order
to be up and running with the following tutorial.

### Ubuntu/Debian

1. Check if the CPU virtualization is enabled.

```
$ grep -o 'vmx\|svm' /proc/cpuinfo
```

2. You may download and install Vagrant from the downloads page [here](https://www.vagrantup.com/downloads).

3. Install a Vagrant provider of your choice — libvirt or VirtualBox.

3a. For libvirt: install KVM, libvirt, QEMU and other dependencies.

```
$ sudo apt update && apt install qemu-kvm libvirt-clients libvirt-daemon-system bridge-utils libguestfs-tools virt-manager
```

3b. You can download and install VirtualBox and VirtualBox Extension Pack from their downloads page [here](https://www.virtualbox.org/wiki/Downloads).

### macOS

1. You can use brew to install VirtualBox and Vagrant.

```
$ brew cask install virtualbox
$ brew cask install vagrant
```

2. The worker will start with UI enabled, the first time it will ask for
   VirtualBox permissions around "Input Monitoring". It is expected just allow
   them.

> ## Project foldering

Reviewer 2: ```suggestion
## Cloning Tinkerbell
```

    provisioner: Configure a credential helper to remove this warning. See
    provisioner: https://docs.docker.com/engine/reference/commandline/login/#credentials-store

    Login Succeeded
    latest: Pulling from tinkerbell/tink-worker
    Digest: sha256:55c99b7cfafd28244fd3f7adbe46d94ada305f0f28553f95dfd241c168f0439e
    Status: Image is up to date for quay.io/tinkerbell/tink-worker:latest
    quay.io/tinkerbell/tink-worker:latest
    The push refers to repository [192.168.1.1/tink-worker]
    c83ceae18bd4: Layer already exists
    a1872780b5d4: Layer already exists
    3e207b409db3: Layer already exists
    latest: digest: sha256:55c99b7cfafd28244fd3f7adbe46d94ada305f0f28553f95dfd241c168f0439e size: 949
    INFO: tinkerbell stack setup completed successfully on ubuntu server
    NEXT: 1. Enter ./deploy and run: source ../envrc; docker-compose up
           2. Try executing your fist workflow.
              Follow the steps described in https://tinkerbell.org/examples/hello-world/ to say 'Hello World!' with a workflow.
```

When the provisioner is ready, a summary is printed to the console and you can
now ssh in.

> ### Connecting with the servers

Reviewer 2: ```suggestion
### Connecting to the Provisioner
```

All the following commands are to be executed being in the artifacts' directory created
above, unless stated otherwise.

Connect to the provisioner via ssh, vagrant has an until:

```
$ vagrant ssh provisioner
vagrant@provisioner:~$
```

At this point you are in an Ubuntu box that has a couple of utils installed and
it is not time to start the services to make it a proper Tinkerbell provisioner

```
$ sudo su -l
$ cd /vagrant && source envrc && cd deploy
$ docker-compose up -d
```

> !TIPS The important sections of the automation script used to bring up the

Reviewer 2: Is this supposed to be a call-out of some sort? Right now it just renders as plain text.

Tinkerbell stack are explained [here](/setup/vagrant/script/).

When you are done with these commands you have a fully working provisioner that
is ready to receive templates and workflow targeting workers. You can check that
running:

```
$ docker-compose ps
        Name                      Command                  State                             Ports
-------------------------------------------------------------------------------------------------------------------------
deploy_boots_1         /boots -dhcp-addr 0.0.0.0: ...   Up
deploy_cacher_1        /cacher                          Up             0.0.0.0:42111->42111/tcp, 0.0.0.0:42112->42112/tcp
deploy_db_1            docker-entrypoint.sh postgres    Up (healthy)   0.0.0.0:5432->5432/tcp
deploy_hegel_1         cmd/hegel                        Up
deploy_nginx_1         /docker-entrypoint.sh ngin ...   Up             192.168.1.2:80->80/tcp
deploy_registry_1      /entrypoint.sh /etc/docker ...   Up (healthy)
deploy_tink-cli_1      /bin/sh -c sleep infinity        Up
deploy_tink-server_1   tink-server                      Up (healthy)   0.0.0.0:42113->42113/tcp, 0.0.0.0:42114->42114/tcp
```

> ### Configure worker hardware

Reviewer 2: ```suggestion
### Register the worker's hardware
```

At this point the provisioner is running and we can follow the [example called
"Hello world"](/examples/hello-world) but I will contextualize it for our vagrant
setup here.

SSH in the provisioner again and we have to define our hello world template and
load it to tinkerbell:

```
$ cat > hello-world.yml
version: "0.1"
name: hello_world_workflow
global_timeout: 600
tasks:
  - name: "hello world"
    worker: "{{.device_1}}"
    actions:
      - name: "hello_world"
        image: hello-world
        timeout: 60
$ docker exec -i deploy_tink-cli_1 tink template create --name hello-world < ./hello-world.yml
Created Template:  75ab8483-6f42-42a9-a80d-a9f6196130df
```

> Now we have to register the future worker (hardware) to tinkerbell

Reviewer 2: ```suggestion
Now we have to register the worker with tinkerbell
```
